src/libsyntax/parse/parser.rs

```diff
@@ -5602,6 +5611,24 @@ impl<'a> Parser<'a> {
             // Parse closing paren and return type.
             self.expect(&token::CloseDelim(token::Paren))?;
         }
+        // Replace duplicated recovered arguments with `_` pattern to avoid unecessary errors.
+        let mut seen_inputs = FxHashSet::default();
+        for input in fn_inputs.iter_mut() {
+            let opt_ident = if let (PatKind::Ident(_, ident, _), ast::ArgSource::Recovery) = (
```

Comment:
Looks like this can be done by checking whether the argument type is `TyKind::Err` without introducing `ArgSource::Recovery`.
(Especially given that `ArgSource` is refactored away in https://github.com/rust-lang/rust/pull/61276.)

Reply:
I'm actually not sure what this logic defends from - `fn foo(u8, u8)`?
Perhaps "arguments without types" are better interpreted as "types without arguments" instead?

Reply:
The problem is we don't know until typeck wether these are types or names, and both are to a first approximation equally likely. Taking the ident and injecting it in the scope as a local is good because we don't emit an error when using the argument without a type. Deduplicating the names is good because that would only likely happen if they were types so we don't emit extraneous irrelevant errors.

I'll do the dedupe using TyErr that makes more sense than what I did.